### PR TITLE
feat: add ECDH-1PU article

### PR DESCRIPTION
**Pull request by baha-ai**
Signed-off-by: Baha Shaaban <baha.shaaban@securekey.com>

### CODE REVIEW

File: docs/spec-files/encryption.md [hunk 1]
```diff
@@ -101,7 +101,7 @@ Assuming she also has Bob's DID document which happens to include the following
 }
 ```
 
-There should be only 1 entry in the recipients of the envelope, representing Bob. The corresponding `kid` header for this recipient MUST have `did:example:jklmnopqrstuvwxyz1#key-1` as value. The packing process MUST extract the public key bytes found in `publicKeyBase58` of Bob's DID Doc `KeyAgreement[0]` to execute the ECDH-1PU key derivation for content key wrapping.
+The recipients of the envelope must include all the KeyAgreement entries, representing Bob. The corresponding `kid` header for this recipient MUST have `did:example:jklmnopqrstuvwxyz1#key-1` as value as per the example above. The packing process MUST extract the public key bytes found in `publicKeyBase58` of Bob's DID Doc `KeyAgreement[0]` to execute the ECDH-1PU key derivation for content key wrapping.
```
Comment (kdenhartog) [inline docs/spec-files/encryption.md:104]:
```suggestion
The recipients of the envelope MUST include all the KeyAgreement entries, representing Bob. The corresponding `kid` header for this recipient MUST have `did:example:jklmnopqrstuvwxyz1#key-1` as value as per the example above. The packing process MUST extract the public key bytes found in `publicKeyBase58` of Bob's DID Doc `KeyAgreement[0]` to execute the ECDH-1PU key derivation for content key wrapping.
```

File: docs/spec-files/ecdh1pu.md [hunk 2]
```diff
@@ -0,0 +1,112 @@
+## ECDH-1PU
```
Comment (kdenhartog) [inline docs/spec-files/ecdh1pu.md:1]: This text seems like it would fit better in the implementation guide rather than the spec.

Comment (Baha-sk) [reply docs/spec-files/ecdh1pu.md:1]: any suggestion where this should be moved?

File: docs/spec-files/encryption.md (hunk 1 above)
Comment (Baha-sk) [reply docs/spec-files/encryption.md]: updated with suggested test @kdenhartog

File: docs/spec-files/encryption.md [hunk 3]
```diff
@@ -101,7 +101,7 @@ Assuming she also has Bob's DID document which happens to include the following
 }
 ```
 
-There should be only 1 entry in the recipients of the envelope, representing Bob. The corresponding `kid` header for this recipient MUST have `did:example:jklmnopqrstuvwxyz1#key-1` as value. The packing process MUST extract the public key bytes found in `publicKeyBase58` of Bob's DID Doc `KeyAgreement[0]` to execute the ECDH-1PU key derivation for content key wrapping.
+The recipients of the envelope MUST include all the KeyAgreement entries, representing Bob. The corresponding `kid` header for this recipient MUST have a DID URL pointing to a verification method in the did document. This verification method MUST be associated with the `KeyAgreement` verification relationship and the public key bytes MUST be retrieved from the DID Document to execute the ECDH-1PU key derivation for content key wrapping.
```
Comment (awoie) [inline docs/spec-files/encryption.md:104]:
```suggestion
The recipients of the envelope MUST include all the KeyAgreement entries, representing Bob. The corresponding `kid` header for this recipient MUST have a DID URL pointing to a verification method in the DID Document. This verification method MUST be associated with the `KeyAgreement` verification relationship and the Verification Material MUST be retrieved from the DID Document to execute the ECDH-1PU key derivation for content key wrapping.
```

Comment (awoie) [inline docs/spec-files/encryption.md:104]: I think it is called Verification Material as per DID Core.

File: docs/spec-files/ecdh1pu.md [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+The former has no sender key involved and is intended for only the recipients of the message. This mechanism requires using encryption with a content encryption key (aka `cek`) being wrapped with a key agreement mechanism using `ECDH-ES` for each recipient. We use key wrapping mode with [ECDH-ES](https://datatracker.ietf.org/doc/html/rfc7518#section-4.6) to ensure only the intended recipients can decrypt the final message. The benefit of using `ECDH-ES` is that it's widely used and available in many crypto libraries for most modern languages. Building JWE envelopes using this type of encryption should be relatively easy using existing JOSE libraries in your preferred language.
+
+The latter however requires a sender key to encrypt the message for the sender to be authenticated. This encryption mechanism uses ECDH-1PU, the main topic of this article. 
+
+### Z in ECDH-ES, first step in understanding ECDH-1PU
+
+To understand how ECDH-1PU, knowledge about the internals of ECDH-ES is required. The way ECDH-ES key agreement works is for the sender to execute the following steps for each recipient to derive the key used to wrap the `cek`:
```
Comment (awoie) [inline docs/spec-files/ecdh1pu.md:15]:
```suggestion
To understand how ECDH-1PU works, knowledge about the internals of ECDH-ES is required. The way ECDH-ES key agreement works is for the sender to execute the following steps for each recipient to derive the key used to wrap the `cek`:
```

File: docs/spec-files/ecdh1pu.md [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+   3. A256CBC-HS512
+   
+    But the spec only mandates [iii](https://identity.foundation/didcomm-messaging/spec/#curves-and-content-encryption-algorithms) as required to minimize interoperability issues in the protocol.
+2. Encrypt the payload prior to wrapping `cek` with the derived `Z`. The output is `ciphertext` and `tag`
+3. Use the resulting `tag` in item 2 above as the value `len(tag)`+`tag` set in `cctag` in the keys derivations of `Ze` and `Zs`.
+
+
+Additionally, the `skid` protected header is also introduced as a kid to reference the sender key. This will help recipients resolve the key behind `skid` and execute the ECDH-1PU process explained in the previous section. 
+
+For the sake of consistency, `apu` and `apv` must be set to the values mentioned in [section 5.8](https://identity.foundation/didcomm-messaging/spec/#ecdh-1pu-key-wrapping-and-common-protected-headers) of the DIDComm Messaging protocol to further restrain and protect the message.
```
Comment (awoie) [inline docs/spec-files/ecdh1pu.md:110]: I think this line is specific to DIDComm. Not sure if this is needed if the article is only about 1PU

**Review by awoie — COMMENTED** [review]
LGTM but perhaps the article could mention how it is been used in DIDComm Messaging. Anyways, great works. We should get Chris to package it as an article. I'm not sure if we want to put this into the DIDComm Messaging spec itself.

File: docs/spec-files/encryption.md (hunk 3 above)
Comment (dhh1128) [reply docs/spec-files/encryption.md:104]: I don't agree with "MUST" here. I think "SHOULD" is more appropriate. We can discuss on a call.

File: docs/spec-files/ecdh1pu.md (hunk 5 above)
Comment (Baha-sk) [reply docs/spec-files/ecdh1pu.md:110]: They're add here only as a hint to refer the spec.

File: docs/spec-files/encryption.md (hunk 3 above)
Comment (Baha-sk) [reply docs/spec-files/encryption.md]: Sure, I'll replace MIUST with SHOILD this this case and add `Unless previously coordinated, the recipients of the ...`

As suggested by @awoie, i'll push a separate PR for this change and revert it here.

File: docs/spec-files/encryption.md (hunk 1 above)
Comment (Baha-sk) [reply docs/spec-files/encryption.md]: i will push a separate PR for this change.. for now, i'm reverting the update in this pr

File: docs/spec-files/ecdh1pu.md (hunk 2 above)
Comment (Baha-sk) [reply docs/spec-files/ecdh1pu.md]: i have moved the file under `docs/guide-files`

File: docs/spec-files/encryption.md (hunk 1 above)
Comment (Baha-sk) [reply docs/spec-files/encryption.md]: new PR to update this sentence: https://github.com/decentralized-identity/didcomm-messaging/pull/270

File: docs/spec-files/encryption.md (hunk 3 above)
Comment (Baha-sk) [reply docs/spec-files/encryption.md]: pr created: https://github.com/decentralized-identity/didcomm-messaging/pull/270

File: docs/spec-files/ecdh1pu.md (hunk 2 above)
Comment (kdenhartog) [reply docs/spec-files/ecdh1pu.md]: 👍 that's where I was thinking they should go. Sorry could have been a bit more clear about that.

**Pull request closed without merging**